pkg/vm/compiler/emit.go

```diff
@@ -1,105 +1,111 @@
 package compiler
 
 import (
-	"bytes"
 	"encoding/binary"
 	"errors"
 	"fmt"
-	"io"
 	"math/big"
 
+	"github.com/CityOfZion/neo-go/pkg/io"
 	"github.com/CityOfZion/neo-go/pkg/util"
 	"github.com/CityOfZion/neo-go/pkg/vm"
 )
 
-func emit(w *bytes.Buffer, instr vm.Instruction, b []byte) error {
-	if err := w.WriteByte(byte(instr)); err != nil {
-		return err
-	}
-	_, err := w.Write(b)
-	return err
+// emit a VM Instruction with data to the given buffer.
+func emit(w *io.BufBinWriter, instr vm.Instruction, b []byte) {
+	emitOpcode(w, instr)
+	w.WriteBytes(b)
 }
 
-func emitOpcode(w io.ByteWriter, instr vm.Instruction) error {
-	return w.WriteByte(byte(instr))
+// emitOpcode emits a single VM Instruction the given buffer.
+func emitOpcode(w *io.BufBinWriter, instr vm.Instruction) {
+	w.WriteBytes(byte(instr))
 }
 
-func emitBool(w io.ByteWriter, ok bool) error {
+// emitBool emits a bool type the given buffer.
+func emitBool(w *io.BufBinWriter, ok bool) {
 	if ok {
-		return emitOpcode(w, vm.PUSHT)
+		emitOpcode(w, vm.PUSHT)
+		return
 	}
-	return emitOpcode(w, vm.PUSHF)
+	emitOpcode(w, vm.PUSHF)
 }
 
-func emitInt(w *bytes.Buffer, i int64) error {
-	if i == -1 {
-		return emitOpcode(w, vm.PUSHM1)
-	}
-	if i == 0 {
-		return emitOpcode(w, vm.PUSHF)
-	}
-	if i > 0 && i < 16 {
+// emitInt emits a int type to the given buffer.
+func emitInt(w *io.BufBinWriter, i int64) {
+	switch {
+	case i == -1:
+		emitOpcode(w, vm.PUSHM1)
+		return
+	case i == 0:
+		emitOpcode(w, vm.PUSHF)
+		return
+	case i > 0 && i < 16:
 		val := vm.Instruction(int(vm.PUSH1) - 1 + int(i))
-		return emitOpcode(w, val)
+		emitOpcode(w, val)
+		return
 	}
 
 	bInt := big.NewInt(i)
 	val := util.ArrayReverse(bInt.Bytes())
-	return emitBytes(w, val)
+	emitBytes(w, val)
 }
 
-func emitString(w *bytes.Buffer, s string) error {
-	return emitBytes(w, []byte(s))
+// emitString emits a string to the given buffer.
+func emitString(w *io.BufBinWriter, s string) {
+	emitBytes(w, []byte(s))
 }
 
-func emitBytes(w *bytes.Buffer, b []byte) error {
-	var (
-		err error
-		n   = len(b)
-	)
+// emitBytes emits a byte array to the given buffer.
+func emitBytes(w *io.BufBinWriter, b []byte) {
+	n := len(b)
 
 	switch {
 	case n <= int(vm.PUSHBYTES75):
-		return emit(w, vm.Instruction(n), b)
+		emit(w, vm.Instruction(n), b)
+		return
 	case n < 0x100:
-		err = emit(w, vm.PUSHDATA1, []byte{byte(n)})
+		emit(w, vm.PUSHDATA1, []byte{byte(n)})
 	case n < 0x10000:
 		buf := make([]byte, 2)
 		binary.LittleEndian.PutUint16(buf, uint16(n))
-		err = emit(w, vm.PUSHDATA2, buf)
+		emit(w, vm.PUSHDATA2, buf)
 	default:
 		buf := make([]byte, 4)
 		binary.LittleEndian.PutUint32(buf, uint32(n))
-		err = emit(w, vm.PUSHDATA4, buf)
-	}
-	if err != nil {
-		return err
+		emit(w, vm.PUSHDATA4, buf)
 	}
-	_, err = w.Write(b)
-	return err
+
+	w.WriteVarBytes(b)
```

Comment:
It will break codegeneration for sure. You need to have the following sequence here: `[PUSHDATAX, len[0], len[1], ..., b[0], b[1], b[2],...]`. After this change you get `[PUSHDATAX, len[0], len[1], ..., len[0], len[1], ..., b[0], b[1], b[2],...]`. The first `len` is from `buf` and the second one is from `WriteVarBytes`.

Reply:
add check for Err back. I guess it was an issue. 
Bad that we don't have unit tests for that: created issue: #534

Reply:
Still relevant.